livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java

```diff
@@ -33,76 +46,81 @@
+  private static Gson gson = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
 
-  // -1 means session is not created yet, valid sessionId start from 0
-  protected int sessionId = -1;
-  protected String appId;
-  protected String webUIAddress;
+  protected SessionInfo sessionInfo;
+  private String livyURL;
+  private long sessionCreationTimeout;
   protected boolean displayAppInfo;
-  protected LivyOutputStream out;
-  protected LivyHelper livyHelper;
 
   public BaseLivyInterprereter(Properties property) {
     super(property);
-    this.out = new LivyOutputStream();
-    this.livyHelper = new LivyHelper(property);
+    this.livyURL = property.getProperty("zeppelin.livy.url");
+    this.sessionCreationTimeout = Long.parseLong(
+        property.getProperty("zeppelin.livy.create.session.timeout", 120 + ""));
   }
 
   public abstract String getSessionKind();
 
   @Override
   public void open() {
-    // TODO(zjffdu) move session creation here.
+    try {
+      initLivySession();
+    } catch (LivyException e) {
+      throw new RuntimeException("Fail to create session, please check livy interpreter log and " +
+          "livy server log", e);
+    }
   }
 
   @Override
   public void close() {
-    if (sessionId != -1) {
-      livyHelper.closeSession(sessionId);
-      // reset sessionId to -1
-      sessionId = -1;
+    if (sessionInfo != null) {
+      closeSession(sessionInfo.id);
+      // reset sessionInfo to null so that we won't close it twice.
+      sessionInfo = null;
     }
   }
 
-  protected void createSession(InterpreterContext context) throws Exception {
-    sessionId = livyHelper.createSession(context, getSessionKind());
+  protected void initLivySession() throws LivyException {
+    this.sessionInfo = createSession(getUserName(), getSessionKind());
     if (displayAppInfo) {
-      this.appId = extractStatementResult(
-          livyHelper.interpret("sc.applicationId", context, sessionId).message().get(0).getData());
-      livyHelper.interpret(
-          "val webui=sc.getClass.getMethod(\"ui\").invoke(sc).asInstanceOf[Some[_]].get",
-          context, sessionId);
-      this.webUIAddress = extractStatementResult(
-          livyHelper.interpret(
-              "webui.getClass.getMethod(\"appUIAddress\").invoke(webui)",
-              context, sessionId).message().get(0).getData());
-      LOGGER.info("Create livy session with sessionId: {}, appId: {}, webUI: {}",
-          sessionId, appId, webUIAddress);
+      if (sessionInfo.appId == null) {
+        // livy 0.2 don't return appId and sparkUiUrl in response so that we need to get it
+        // explicitly by ourselves.
+        sessionInfo.appId = extractStatementResult(
+            interpret("sc.applicationId", false).message()
+                .get(0).getData());
+      }
+
+      interpret(
+          "val webui=sc.getClass.getMethod(\"ui\").invoke(sc).asInstanceOf[Some[_]].get", false);
```

Comment:
isn't this `uiWebUrl`, instead of `ui`?
also, older Spark doesn't have this, would it cause an exception to call `invoke`?